scintillometry/base/base.py

```diff
@@ -0,0 +1,195 @@
+        """Current offset in the file.
+
+        Parameters
+        ----------
+        unit : `~astropy.units.Unit` or str, optional
+            Time unit the offset should be returned in.  By default, no unit
+            is used, i.e., an integer enumerating samples is returned. For the
+            special string 'time', the absolute time is calculated.
+
+        Returns
+        -------
+        offset : int, `~astropy.units.Quantity`, or `~astropy.time.Time`
+             Offset in current file (or time at current position).
+        """
+        if unit is None:
+            return self.offset
+
+        if unit == 'time':
+            return self.start_time + self.tell(unit=u.s)
+
+        return (self.offset / self.sample_rate).to(unit)
+
+    def read(self, count=None, out=None):
+        """Read a number of complete samples.
+
+        Parameters
+        ----------
+        count : int or None, optional
+            Number of complete samples to read.  If `None` (default) or
+            negative, the entire input data is processed.  Ignored if ``out``
+            is given.
+        out : None or array, optional
+            Array to store the output in. If given, ``count`` will be inferred
+            from the first dimension; the other dimension should equal
+            `sample_shape`.
+
+        Returns
+        -------
+        out : `~numpy.ndarray` of float or complex
+            The first dimension is sample-time, and the remainder given by
+            `sample_shape`.
+        """
+        # NOTE: this will return an EOF error when attempting to read partial
+        # frames, making it identical to fh.read().
+
+        if out is None:
+            if count is None or count < 0:
+                count = self.shape[0] - self.offset
+                if count < 0:
+                    raise EOFError("cannot read from beyond end of input.")
+            out = np.empty((count,) + self.shape[1:], dtype=self.dtype)
+        else:
+            assert out.shape[1:] == self.shape[1:], (
+                "'out' should have trailing shape {}".format(self.sample_shape))
+            count = out.shape[0]
+
+        offset0 = self.offset
+        sample = 0
+        while count > 0:
+            # For current position, get frame plus offset in that frame.
+            frame_index, sample_offset = divmod(self.offset,
+                                                self._samples_per_frame)
+
+            if frame_index != self._frame_index:
+                self._frame = self._read_frame(frame_index)
+                self._frame_index = frame_index
+
+            nsample = min(count, len(self._frame) - sample_offset)
+            data = self._frame[sample_offset:sample_offset + nsample]
+            # Copy to relevant part of output.
+            out[sample:sample + nsample] = data
+            sample += nsample
+            self.offset = offset0 + sample
+            count -= nsample
+
+        return out
+
+    def close(self):
```

Comment:
Should we have `__enter__` and `__exit__` methods as well? I considered `closed` as well, but we do not have that in baseband either, so probably fine to omit.

Reply:
I'm not a huge fan of writing something like:
```
with channelize(dedisperse(vdif.open('myfile.vdif'))) as ch:
    ...
```
It would work, but it's really ugly and I'd rather not the user get acquainted with writing things this way.  I almost think the `__enter__` and `__exit__` should be part of the task governor that uses sciluigi or something to manage the tasks.

Reply:
OK, raised new issue to remind us (me...)